w4/domain/usecase/merchant_usecase.go

```diff
@@ -0,0 +1,104 @@
+package usecase
+
+import (
+	"belimang/domain/entity"
+	"belimang/domain/repository"
+	"belimang/protocol/api/dto"
+)
+
+type IMerchantUsecase interface {
+	Create(payload MerchantCreatePayload) (dto.MerchantCreateDtoResponse, error)
+	Fetch(query MerchantFetchQuery) ([]dto.MerchantFetchDtoResponse, error)
+}
+
+type merchantUsecase struct {
+	merchantRepository repository.IMerchantRepository
+}
+
+func NewMerchantUsecase(merchantRepository repository.IMerchantRepository) *merchantUsecase {
+	return &merchantUsecase{merchantRepository}
+}
+
+type MerchantCreatePayload struct {
+	Name     string                  `json:"name" validate:"required,min=2,max=30"`
+	Category entity.MerchantCategory `json:"merchantCategory" validate:"required,category=merchant"`
+	ImageUrl string                  `json:"imageUrl" validate:"required,urlformat"`
+	Location entity.Location         `json:"location" validate:"required"`
+}
+
+func (u *merchantUsecase) Create(payload MerchantCreatePayload) (dto.MerchantCreateDtoResponse, error) {
+	var entity entity.Merchant
+
+	entity.Name = payload.Name
+	entity.Category = payload.Category
+	entity.ImageUrl = payload.ImageUrl
+	entity.Lat = payload.Location.Lat
+	entity.Lon = payload.Location.Lon
+	entity, err := u.merchantRepository.Create(entity)
+	if err != nil {
+		return dto.MerchantCreateDtoResponse{}, err
+	}
+
+	return dto.MerchantCreateDtoResponse{ID: entity.ID.String()}, nil
+}
+
+type MerchantFetchQuery struct {
+	ID            string                  `query:"merchantId"`
+	Name          string                  `query:"name"`
+	Category      entity.MerchantCategory `query:"merchantCategory"`
+	ImageUrl      string                  `query:"imageUrl"`
+	Location      entity.Location         `query:"location"`
+	SortCreatedAt entity.SortType         `query:"createdAt"`
+	Limit         int                     `query:"limit"`
+	Offset        int                     `query:"offset"`
+}
+
+func (u *merchantUsecase) Fetch(query MerchantFetchQuery) ([]dto.MerchantFetchDtoResponse, error) {
+	var (
+		response []dto.MerchantFetchDtoResponse = make([]dto.MerchantFetchDtoResponse, 0)
+		filter   entity.MerchantFetchFilter
+	)
+
+	if query.ID != "" {
+		filter.ID = query.ID
+	}
+
+	if query.Name != "" {
+		filter.Name = query.Name
+	}
+
+	if query.Category.String() != "" {
+		if query.Category.Valid() {
+			filter.MerchantCategory = query.Category
+		} else {
+			return response, nil
+		}
+	}
```

Comment:
It can be done without nested loop right ? 

```json
if !query.Category.Valid() return response, nil
otherwise set the value 
```

Reply:
There's a reason as to why I do it that why. As stated in the requirements below
![image](https://github.com/shamir92/project-sprint-team-happy/assets/25399846/34e88363-aec0-4218-9a01-62d2bc5104f9)

If we only validate the category provided is invalid, we will always get empty result, because `empty string` is considered **invalid category**, that's why we need to check if the `merchantCategory` is null/empty before validating its existence.